### ISSUE TITLE
[meta] Vulkan presentation on Nvidia Optimus/Prime

### ISSUE THREAD

**Issue by kvark** [kvark/blade#205]
Trying to collect everything we know about this annoying blocker.
TL;DR: on Intel+Nvidia systems, under unknown circumstances, one of the physical devices is failing to either create a surface or present on it, despite Vulkan listing it as the first and specifying the ability to present.

Upstream:
- https://gitlab.freedesktop.org/mesa/mesa/-/issues/4688
- https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/10430

Issues:
- https://github.com/kvark/blade/issues/88
- https://github.com/gfx-rs/wgpu-rs/issues/673

Mentions:
- https://github.com/zed-industries/zed/pull/20853#issuecomment-2490257411

Affected user configurations:
| DE | Nvidia | Intel | source | notes |
| -- | ------- | ----- | ------ | ------ | 
| X11 | 460.67 | **?** | https://gitlab.freedesktop.org/mesa/mesa/-/issues/4688 | |
| X11 | 470.223.2.0 | **23.2.1** | https://github.com/kvark/blade/issues/88#issue-2149930187 | |
| X11 | 1975239040 | **96477185** | https://github.com/zed-industries/zed/issues/8168#issuecomment-2103724277 | Nvidia is first, but lacks "KHR_dynamic_rendering" |
| X11 | 470.239 | **?** | https://github.com/zed-industries/zed/issues/8168#issuecomment-2109177630 | |
| Wayland | **525.105.17** | ? | https://github.com/NVIDIA/open-gpu-kernel-modules/issues/317#issuecomment-1533978313| |
| Wayland | **530.41.03** | ? | https://github.com/mpv-player/mpv/issues/11711#issue-1722157407 | |
| Wayland | **550.78.0.0** | 24.0.6 | https://github.com/zed-industries/zed/issues/11475#issue-2282126071 | |
| Wayland | **550.54.14.0** | N/A | https://github.com/zed-industries/zed/issues/8168#issuecomment-1985867956 |  single-GPU system |

Ideas:
- there appears to be a correlation with platforms. Users on X11 are unable to present with Intel (when Prime selects Nvidia), but users on Wayland are unable to present with Nvidia.

**Comment by Blaeriz** [comment]
### Zed/Blade fails

## Hardware

```bash
MSI Katana 15 B13VEK (REV:1.0)
13th Gen Intel(R) Core(TM) i7-13620H (16) @ 4.90 GHz
NVIDIA GeForce RTX 4050 Max-Q / Mobile [Discrete]
Intel UHD Graphics @ 1.50 GHz [Integrated]
```

## Zed error

```bash
2025-05-01T22:42:39+05:30 WARN  [blade_graphics::hal::surface] Rejected for not presenting to the window surface
2025-05-01T22:42:39+05:30 INFO  [*unknown*] read_socket;
2025-05-01T22:42:39+05:30 INFO  [*unknown*] read_socket;
2025-05-01T22:42:39+05:30 ERROR [zed::reliability] {
  "thread": "main",
  "payload": "called `Result::unwrap()` on an `Err` value: PlatformNotSupported",
  "location_data": {
    "file": "crates/gpui/src/platform/blade/blade_renderer.rs",
    "line": 345
  },
  "backtrace": [
    "__libc_start_main+28632016848745"
  ],
  "app_version": "0.184.8",
  "release_channel": "stable",
  "target": "x86_64-unknown-linux-gnu",
  "os_name": "Linux Wayland",
  "os_version": "arch unknown",
  "architecture": "x86_64",
  "panicked_on": 1746119559419,
  "system_id": "f708bdd3-e126-4c85-98fd-0a49fa2b1518",
  "installation_id": "2a07361d-bf80-4b0a-812b-ce3626cc6dca",
  "session_id": "9e01d921-ba0e-49de-be24-14fc8a7a9d66"
}
```

## VKcube behavior

```bash kesh : ~ > $ vkcube --wsi wayland --gpu_number 0 
Selected GPU 0: NVIDIA GeForce RTX 4050 Laptop GPU, type: DiscreteGpu Could not find both graphics and present queues ``` 

## Driver Versions

```bash
 kesh : ~ > $ vulkaninfo | grep driverVersion
    driverVersion     = 570.144.0.0 (2393112576)
    driverVersion     = 25.0.4 (104857604)
```
